opm/parser/eclipse/EclipseState/Schedule/ActionAST.hpp

```diff
@@ -0,0 +1,135 @@
+/*
+  Copyright 2018 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef ActionAST_HPP
+#define ActionAST_HPP
+
+#include <string>
+#include <vector>
+
+
+namespace Opm {
+
+class ActionContext;
+
+enum TokenType {
+    number,        //  0
+    ecl_expr,      //  1
+    open_paren,    //  2
+    close_paren,   //  3
+    op_gt,         //  4
+    op_ge,         //  5
+    op_lt,         //  6
+    op_le,         //  7
+    op_eq,         //  8
+    op_ne,         //  9
+    op_and,        // 10
+    op_or,         // 11
+    end,           // 12
+    error          // 13
+};
+
+struct ParseNode {
+    ParseNode(TokenType type, const std::string& value) :
+        type(type),
+        value(value)
+    {}
+
+
+    ParseNode(TokenType type) : ParseNode(type, "")
+    {}
+
+
+    TokenType type;
+    std::string value;
+};
+
+
+class ASTNode {
+public:
+
+ASTNode() :
+    type(TokenType::error)
+{}
+
+
+ASTNode(TokenType type):
+    type(type)
+{}
+
+
+ASTNode(double value) :
+    type(TokenType::number),
+    number(value)
+{}
+
+
+ASTNode(TokenType type, const std::string& func, const std::vector<std::string>& arg_list):
+    type(type),
+    func(func),
+    arg_list(arg_list)
+{}
+
+    bool eval(const ActionContext& context) const;
+    double value(const ActionContext& context) const;
+    TokenType type;
+    void add_child(const ASTNode& child);
+    size_t size() const;
+
+private:
+    std::string func;
+    std::vector<std::string> arg_list;
+    double number;
+    std::vector<ASTNode> children;
```

Comment:
Do note that having a `vector<T>` as a member of `T` (in this case, `T = ASTNode`) does to a certain extent depend on undefined behaviour&mdash;at least until C++17 which explicitly blesses this particular usage.  It depends on being able to instantiate objects of type `std::allocator<T>` when `T` is an incomplete type and, subtly, on `vector<>` being implemented in terms of `T*`.

You may want to add a comment to that effect here.

Reply:
OK - I'll do that; I must admit these C++ peculiarities annoy me a bit. If it compiles without warnings (for some reasonably high warning level); and works - I would like to think the code was decent from a standards point of view.

Reply:
> I must admit these C++ peculiarities annoy me a bit. If it compiles without warnings (for some reasonably high warning level); and works - I would like to think the code was decent from a standards point of view.

I appreciate the sentiment.  To be fair, with libstdc++ the compiler only sees something that looks a lot like
```C++
class ASTNode
{
// ...
private:
    class V
    {
    // ...
    private:
        ASTNode* __x;
    // ...
    };
// ...
    V children;
};
```
and this is completely fine from a language perspective and therefore no diagnostic is required.